### ISSUE TITLE
Issues with / (over)

### ISSUE THREAD

**Issue by kirbyfan64** [kevinlawler/kona#380]
Not sure what exactly is going on here, but:

```
ryan@DevPC-LX:~$ rlwrap k2
K 2.8 2000-10-10 Copyright (C) 1993-2000 Kx Systems
Evaluation. Not for commercial use. 
\ for help. \\ to exit.

  -:/1 / loops forever (or until I hit ^c)
stop
-:/1 / loops forever (or until I hit ^c)
^
>  
ryan@DevPC-LX:~$ rlwrap k
K Console - Enter \ for help

  -:/1 / returns immediately
1
  
ryan@DevPC-LX:~$ 
```

**Comment by tavmem** [comment]
Not clear (yet) whether this is a bug in k2.8 or in Kona (probably in Kona).

Some other differences related to this
in k2.8 
```
  -:1
syntax error
```
In Kona
```
  -:1
-1
```
In k2.8
```
  a:5
  a-:/1
-1
  a
5
```
in Kona
```
  a:5
5
  a-:/1
nyi error
>  a
5
> 
```

**Comment by tavmem** [comment]
Note how k4 works
```
q)\
  -:/1
-1
  -:1
-1
  a:5
  a-:/1
-1
  a
5
```
In the 1st 2 cases, Kona works exactly like k4
In the last case, k4 works like k2.8, and Kona yields "nyi error".

This lends some credibility to the thesis that the differences in the 1st 2 cases are due to bugs in k2.8, and that the 3rd case is simply a missing feature in Kona.

**Comment by kirbyfan64** [comment]
@tavmem Isn't `/` given a monadic verb supposed to loop until the value stops changing, though?

**Comment by tavmem** [comment]
That's a point ...
so then, give me a case that yields a useable result in k2.8
i.e., does not loop forever.

**Comment by tavmem** [comment]
Unless you mean that it's supposed to loop forever.

**Comment by kirbyfan64** [comment]
@tavmem I think it *is* supposed to loop forever. k4 (and seemingly k5) changed the behavior to mean "keep looping until a value that has been seen before shows up", while k2 meant "keep looping until the same value is returned twice in a row".

**Comment by kirbyfan64** [comment]
In addition, Kona doesn't even return the correct result; given `-:/1`, k5 and k4 BOTH return `-1`, while Kona returns `1`:

```
ryan@DevPC-LX:~/rejit2$ rlwrap k
K Console - Enter \ for help

  {`0:,$x;-x}/1
1
-1
1
  {`0:,$x;-x}\1
1
-1
1 -1
  
```

In addition, Kona follows k2 in some places; this:

```
i::5;j::5;{if[~i;if[j;j-:x];:j];i-:x;i}/1
```

will result in both i and j becoming 0 with the k2 behavior and j staying at 5 with the k4/k5 behavior. Kona follows k2 here.

**Comment by tavmem** [comment]
Good points.
I think that as a general rule Kona should mimic the behavior of k2.8 and k3.2, unless there is a really obvious bug in k2.8 and k3.2.
In particular, any scripts that work in k2.8 or k3.2 should work in Kona.
I will tag this a a Kona bug.

**Comment by tavmem** [comment]
Actually, it's beginning to look like a bug both in k2.8 and in Kona
The definition in kref for Over Monad for the case f/ x
is "Apply f interatively to x until the result matches either the previous or the initial result."
```
  f: -:
-:
  f 1
-1
  f -1
1
  f 1
-1
```
The initial result is -1, which recurs as the 3rd result.
Based on the definition, k2.8 should not keep looping forever.
But it seems that the final result of f/1 should be -1.

**Comment by tavmem** [comment]
@kirbyfan64 wrote:
>  I think it is supposed to loop forever. k4 (and seemingly k5) changed the behavior to mean "keep looping until a value that has been seen before shows up",

Where did you get the quote on the k4 definition?
In ***q for Mortals*** (2008 edition) the definition for "over" (page 122)seems to exclude the dyadic case:
"The over adverb / modifies a base dyadic function ..."

However, the monadic verb ***neg*** yields the following
```
q)(neg/)1
-1
```
which is consistent with the definition of monadic over given in kref.

**Comment by tavmem** [comment]
Another example: take the k2.8  monadic function.  k2.8 iterates endlessly
```
  f:{(x+1)!3}
  f/2
stop
{(x+1)!3}
       ^
> 
```
However, f cycles
```
  f 2
0
  f 0
1
  f 1
2
  f 2
0
```
So, it appears from the definition in kref that f/2 should be 0.
However, Kona gives these results
```
  f:{(x+1)!3}
{(x+1)!3}
  f/2
2

  -:/ 1
1

```
Kona yields the x which will give the initial result (not the initial result).
In q
```
q)f:{(x+1)mod 3}
q)(f/)2
1

q)f 2
0
q)f 0
1
q)f 1
2

q)(neg/)1
-1

q)neg 1
-1
q)neg -1
1
```
q yields the value that yields the initial x

**Comment by tavmem** [comment]
Apparent conclusions:
Kona follows **an interpretation** of the literal definition listed in kref.
Q also follows **an interpretation** of the kref definition, but a different interpretation, which is not documented in **q for Mortals**.
k3.2 has a bug.

Defintion in kref
> Apply f interatively to x until, in the first case, a result matches either the previous or the initial result, ...
Notably, the definition says **a result**, not **the result**.

Most importantly, I just noticed that the k2.8 bug was fixed in k3.2
```
  -:/1
-1

  f:{(x+1)!2}
  f/2
1
```
The results in k3.2 and in q are exactly the same.
It seems that Kona should match the k3.2 results.
